xl2times/transforms.py

```diff
@@ -570,6 +570,69 @@ def process_user_constraint_table(
     return [process_user_constraint_table(t) for t in tables]
 
 
+def generate_uc_properties(
+    config: datatypes.Config,
+    tables: List[datatypes.EmbeddedXlTable],
+    model: datatypes.TimesModel,
+) -> List[datatypes.EmbeddedXlTable]:
+    """
+    Generate a dataframe containing User Constraint properties
+    """
+
+    uc_tables = [table for table in tables if table.tag == datatypes.Tag.uc_t]
+    columns = ["uc_n", "uc_desc", "region", "reg_action", "ts_action"]
+    user_constraints = pd.DataFrame(columns=columns)
+
+    for uc_table in uc_tables:
+        df = uc_table.dataframe.loc[:, ["uc_n"]].drop_duplicates(keep="first")
+        df = df.merge(
+            uc_table.dataframe.loc[:, ["uc_n", "uc_desc"]]
+            .drop_duplicates(keep="first")
```

Comment:
No, because of `.dropna(), how="left")` at the end. :) 
Here we suplement a column of user constraint names with their descriptions (if present). Later on, missing descriptions are populated using user constraint names.
I can add a comment to explain this.

Reply:
It still feels redundant. Why not replace both the lines starting `df = ` with:
```python
df = uc_table.dataframe.loc[:, ["uc_n", "uc_desc"]].drop_duplicates(subset='uc_n', keep='first')
```

Reply:
Actually, perhaps this way is better (will also handle NAs):
```python
In [3]: df2 = pd.DataFrame({'uc_n': ['A', 'B', 'C', 'B', 'C'], 'uc_desc': ['Ad',
   ...: None, 'Cd', 'Bd', 'Cd1']})

In [4]: df2
Out[4]: 
  uc_n uc_desc
0    A      Ad
1    B    None
2    C      Cd
3    B      Bd
4    C     Cd1

In [5]: df2.groupby('uc_n', sort=False).first()
Out[5]: 
     uc_desc
uc_n        
A         Ad
B         Bd
C         Cd
```

Reply:
> It still feels redundant. Why not replace both the lines starting `df = ` with:
> 
> ```python
> df = uc_table.dataframe.loc[:, ["uc_n", "uc_desc"]].drop_duplicates(subset='uc_n', keep='first')
> ```

I know! It would introduce a regression though. This replaces the code that was introduced in #118 that was used to supplement similar approach.

Reply:
> ```python
> df2 = pd.DataFrame({'uc_n': ['A', 'B', 'C', 'B', 'C'], 'uc_desc': ['Ad',
>    ...: None, 'Cd', 'Bd', 'Cd1']})
> ```

Ok, this may actually work. Will try it, thanks!

Reply:
Actually, there are is depreciation warning for both `DataFrame.first()` and `DataFrame.last()`: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.first.html

Will keep it as is for now!

Reply:
We aren't using DataFrame.first(), we're using an identically named method from GroupBy:
https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.DataFrameGroupBy.first.html#pandas.core.groupby.DataFrameGroupBy.first

I would recommend again to use the groupby approach, it is around 5x faster than the current code:
```python
In [19]: %%timeit
    ...: df = df2.loc[:, ["uc_n"]].drop_duplicates(keep="first")
    ...: df = df.merge(
    ...:     df2.loc[:, ["uc_n", "uc_desc"]]
    ...:     .drop_duplicates(keep="first")
    ...:     .dropna(),
    ...:     how='left',
    ...:     )
    ...: 
2.57 ms ± 62.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [20]: %timeit df2.groupby('uc_n', sort=False).first()
414 µs ± 7.53 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

Reply:
There are probably other places where we can use it as well. Will open an issue to investigate and make a consitent change accross the tool.